cf-agent/package_module.c

```diff
@@ -0,0 +1,1604 @@
+                                        Rlist **response,
+                                        const PackageModuleWrapper *wrapper)
+{
+    assert(args && wrapper);
+    
+    char *command = StringFormat("%s %s", wrapper->path, args);
+    IOData io = cf_popen_full_duplex(command, false);
+    free(command);
+    
+    if (io.write_fd == 0 || io.read_fd == 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating "
+            "package module.");
+        return -1;
+    }
+    
+    if (WriteScriptData(request, &io) != strlen(request))
+    {
+        Log(LOG_LEVEL_VERBOSE, "Couldn't send whole data to package module.");
+        return -1;
+    }
+    
+    /* We can have some error message here. */
+    Rlist *res = ReadDataFromPackageScript(&io);
+    
+    /* If script returns non 0 status */
+    if (cf_pclose_full_duplex(&io) != EXIT_SUCCESS)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Package module returned with non zero return code.");
+        RlistDestroy(res);
+        return -1;
+    }
+    
+    *response = res;
+    return 0;
+}
+
+static void FreePackageInfo(PackageInfo *package_info)
+{
+    if (package_info)
+    {
+        free(package_info->arch);
+        free(package_info->name);
+        free(package_info->version);
+
+        free(package_info);
+    }
+}
+
+static char *ParseOptions(Rlist *options)
+{
+    if (RlistIsNullList(options))
+    {
+        return SafeStringDuplicate("");
+    }
+    
+    Buffer *data = BufferNew();
+    for (Rlist *rp = options; rp != NULL; rp = rp->next)
+    {
+        char *value = RlistScalarValue(rp);
+        BufferAppendString(data, "options=");
+        BufferAppendString(data, value);
+        BufferAppendString(data, "\n");
+    }
+    return BufferClose(data);
+}
+
+static PackageInfo *ParseAndCheckPackageDataReply(const Rlist *data)
+{
+    PackageInfo * package_data = xcalloc(1, sizeof(PackageInfo));
+    
+    for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+    {
+        char *line = RlistScalarValue(rp);
+                   
+        if (StringStartsWith(line, "PackageType="))
+        {
+            char *type = line + strlen("PackageType=");
+            if (StringSafeEqual(type, "file"))
+            {
+                package_data->type = PACKAGE_TYPE_FILE;
+            }
+            else if (StringSafeEqual(type, "repo"))
+            {
+                package_data->type = PACKAGE_TYPE_REPO;
+            }
+            else
+            {
+                Log(LOG_LEVEL_VERBOSE, "unsupported package type: %s", type);
+                free(package_data);
+                return NULL;
+            }
+        }
+        else if (StringStartsWith(line, "Name="))
+        {
+            if (package_data->name)
+            {
+                /* Some error occurred as we already have name for 
+                 * given package. */
+                Log(LOG_LEVEL_VERBOSE, "duplicated package name "
```

Comment:
I would make all unexpected scenarios, especially with regards to the protocol communication, into ``LOG_LEVEL_ERR``. They are indicative of a bug in either the code or the package module, and should stand out.